platformics/security/authorization.py

```diff
@@ -2,44 +2,142 @@
+
+    if "project_roles" not in claims:
+        raise Exception("no project roles in claims")
+
+    project_claims = claims["project_roles"]
+
+    try:
+        for role, project_ids in project_claims.items():
+            assert role in ["member", "owner", "viewer"]
+            assert isinstance(project_ids, list)
+            for item in project_ids:
+                assert int(item)
+    except Exception:
+        return None
+
+    return Principal(
+        claims["sub"],
+        roles=["user"],
+        attr={
+            "user_id": int(claims["sub"]),
+            "owner_projects": project_claims.get("owner", []),
+            "member_projects": project_claims.get("member", []),
+            "viewer_projects": project_claims.get("viewer", []),
+            "service_identity": claims["service_identity"],
+        },
     )
-    return query
+
+
+class AuthzClient:
+    def __init__(self, settings: APISettings):
+        self.settings = settings
+        self.client = CerbosClient(host=settings.CERBOS_URL)
+
+    # Convert a model object to a dictionary
+    def _obj_to_dict(self, obj):
+        mydict = {}
+        relationships = obj.__mapper__.relationships
+        for col in obj.__mapper__.all_orm_descriptors:
+            # Don't send related fields to cerbos for authz checks
+            if col.key in relationships:
+                continue
+            value = getattr(obj, col.key)
+            if type(value) not in [int, str, bool, float]:
+                # TODO, we probably want to look into a smarter way to serialize fields for cerbos
+                value = str(value)
+            mydict[col.key] = value
+        return mydict
+
+    def can_create(self, resource, principal: Principal) -> bool:
+        resource_type = type(resource).__tablename__
+        attr = self._obj_to_dict(resource)
+        resource = Resource(id="NEW_ID", kind=resource_type, attr=attr)
+        if self.client.is_allowed(AuthzAction.CREATE, principal, resource):
+            return True
+        return False
+
+    def can_update(self, resource, principal: Principal) -> bool:
+        resource_type = type(resource).__tablename__
+        attr = self._obj_to_dict(resource)
+        # TODO - this should send in the actual resource ID instead of a placeholder string
+        # There are two complexities there though: UUID's don't natively serialize to json,
+        # so they cannot be sent in cerbos perms checks, and we need to find/use the table's
+        # primary key instead of a hardcoded column name.
+        resource = Resource(id="resource_id", kind=resource_type, attr=attr)
+        if self.client.is_allowed(AuthzAction.UPDATE, principal, resource):
+            return True
+        return False
+
+    # Get a SQLAlchemy model with authz filters already applied
+    def get_resource_query(
+        self,
+        principal: Principal,
+        action: AuthzAction,
+        model_cls: type[db.Base],  # type: ignore
+    ) -> Select:
```

Comment:
how overridable do you think this is if we're not using CERBOS?

Reply:
It's TBD honestly. This hopefully get us one step closer by putting most of the policy enforcement code in one place and allowing it to be overridden, but I'd love to hear more about what the interface for the authz systems you've been working with recently look like!